src/test/unit/io/deserializer_free_test.cpp

```diff
@@ -0,0 +1,459 @@
+#include <stan/io/deserializer.hpp>
+// expect_near_rel comes from lib/stan_math
+#include <test/unit/math/expect_near_rel.hpp>
+#include <gtest/gtest.h>
+
+// lb
+
+TEST(deserializer_vector, lb_free) {
+  double lb = 0.5;
+  Eigen::VectorXd uvec1 = Eigen::VectorXd::Random(3);
+  Eigen::VectorXd uvec2 = Eigen::VectorXd::Random(3);
+  Eigen::VectorXd vec_bad(2);
+  Eigen::VectorXd vec1 = stan::math::lb_constrain(uvec1, lb);
+  Eigen::VectorXd vec2 = stan::math::lb_constrain(uvec2, lb);
+  vec_bad << -0.5, 1.0;
+
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  std::vector<Eigen::VectorXd> std_vec = {vec1, vec2},
+                               std_vec_bad = {vec1, vec_bad};
+
+  std::vector<std::vector<Eigen::VectorXd>> std_std_vec
+      = {{vec1, vec2}, {vec2, vec1}};
+
+  stan::test::expect_near_rel("deserializer free", uvec1,
+                              deserializer.free_lb(vec1, lb));
+  std::vector<Eigen::VectorXd> std_uvec = deserializer.free_lb(std_vec, lb);
+  stan::test::expect_near_rel("deserializer free", uvec1, std_uvec[0]);
+  stan::test::expect_near_rel("deserializer free", uvec2, std_uvec[1]);
+  std::vector<std::vector<Eigen::VectorXd>> std_std_uvec
+      = deserializer.free_lb(std_std_vec, lb);
+  stan::test::expect_near_rel("deserializer free", uvec1, std_std_uvec[0][0]);
+  stan::test::expect_near_rel("deserializer free", uvec2, std_std_uvec[0][1]);
+  stan::test::expect_near_rel("deserializer free", uvec1, std_std_uvec[1][1]);
+  stan::test::expect_near_rel("deserializer free", uvec2, std_std_uvec[1][0]);
+
+  EXPECT_THROW(deserializer.free_lb(vec_bad, lb), std::domain_error);
+  EXPECT_THROW(deserializer.free_lb(std_vec_bad, lb), std::domain_error);
+}
+
+// ub
+
+TEST(deserializer_vector, ub_free) {
+  double ub = 0.5;
+  Eigen::VectorXd uvec1 = Eigen::VectorXd::Random(3);
+  Eigen::VectorXd uvec2 = Eigen::VectorXd::Random(3);
+  Eigen::VectorXd vec_bad(2);
+  Eigen::VectorXd vec1 = stan::math::ub_constrain(uvec1, ub);
+  Eigen::VectorXd vec2 = stan::math::ub_constrain(uvec2, ub);
+  vec_bad << -0.5, 1.0;
+
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+
+  std::vector<Eigen::VectorXd> std_vec = {vec1, vec2},
+                               std_vec_bad = {vec1, vec_bad};
+
+  std::vector<std::vector<Eigen::VectorXd>> std_std_vec
+      = {{vec1, vec2}, {vec2, vec1}};
```

Comment:
Would have saved us a lot of code if I had remembered that std::vector versions just call the underlying and could have done them all in one go. Very nice!